encoding/prefix.go

```diff
@@ -17,7 +17,10 @@ var (
 func ReadPrefix(r io.Reader, maxLen uint64) ([]byte, error) {
 	prefix := make([]byte, 8)
 	if n, err := io.ReadFull(r, prefix); n == 0 {
-		return nil, errNoData
+		if err == nil {
+			err = errNoData
+		}
+		return nil, err
```

Comment:
it's actually related to the blacklist haha. The blacklist relies on getting timeout errors, but if a timeout error prevents you from reading any data, you'll get `no data` instead of `timeout`. I'm not sure why I thought this was a great idea in the first place, honestly